### PR TITLE
LPS-86352 Support Checkbox Multiple, Grid and Repeatable fields on Field model

### PR DESCRIPTION
**Pull request by phcp**
Guys, this is part of our forms screenlet, I'll split in some PRs to avoid tons of files to review.

@dgarciasarai @victorg1991 @nhpatt

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,7 @@ All notable changes to this project will be documented in this file.
 * Configure `detekt` for code convention and format all the project
 * Update ThingScreenlet with ApioConsumer 0.0.8 changes
 * Support workflow list with ThingScreenlet
+* Support Checkbox Multiple, Grid and Repeatable fields on Field model
```
Comment (nhpatt) [inline CHANGELOG.md]: 👏 👏 👏

File: android/library/src/main/java/com/liferay/mobile/screens/ddl/model/DDMStructure.java [hunk 2]
```diff
@@ -60,10 +64,33 @@ public DDMStructure createFromParcel(Parcel in) {
     private Long classNameId;
     private String classPK;
 
+    private List<FormPage> pages;
```
Comment (nhpatt) [inline android/library/src/main/java/com/liferay/mobile/screens/ddl/model/DDMStructure.java]: why not new ArrayList to not have to check for null values?

File: android/library/src/main/java/com/liferay/mobile/screens/ddl/model/DDMStructure.java [hunk 3]
```diff
@@ -73,6 +100,8 @@ protected DDMStructure(Parcel in, ClassLoader classLoader) {
         Parcelable[] array = in.readParcelableArray(Field.class.getClassLoader());
         fields = new ArrayList(Arrays.asList(array));
         locale = (Locale) in.readSerializable();
+        name = in.readString();
+        description = in.readString();
```
Comment (nhpatt) [inline android/library/src/main/java/com/liferay/mobile/screens/ddl/model/DDMStructure.java]: new fields should be added here, right?

File: android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/CheckboxMultipleField.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            public SelectableOptionsField createFromParcel(Parcel in) {
+                throw new AssertionError();
+            }
+
+            public SelectableOptionsField[] newArray(int size) {
+                return new CheckboxMultipleField[size];
+            }
+        };
+
+    private boolean isShowAsSwitcher;
```
Comment (nhpatt) [inline android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/CheckboxMultipleField.java]: in this case the "is" is very ugly

Comment (nhpatt) [reply android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/CheckboxMultipleField.java]: Or shown

Comment (phcp) [reply android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/CheckboxMultipleField.java]: That's true, really bad. I'll remove it.

File: android/library/src/main/java/com/liferay/mobile/screens/ddl/model/DDMStructure.java (hunk 2 above)
Comment (phcp) [reply android/library/src/main/java/com/liferay/mobile/screens/ddl/model/DDMStructure.java]: Good point, I will add it.

File: android/library/src/main/java/com/liferay/mobile/screens/ddl/model/DDMStructure.java (hunk 3 above)
Comment (phcp) [reply android/library/src/main/java/com/liferay/mobile/screens/ddl/model/DDMStructure.java]: Yep, I fixed the parcelable implementation in the last commits.

**Comment by nhpatt** [comment]
LGTM, can you @victorg1991 do a quick review **tomorrow** and merge it?

**Comment by dgarciasarai** [comment]
Review travis errors :)

**Comment by nhpatt** [comment]
Yeah, I noticed that Travis was failing due to 65k methods, I don't know if it's solvable before merging :(

File: android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/FormContext.kt [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+import com.liferay.apio.consumer.model.Thing
+import com.liferay.apio.consumer.model.get
+
+/**
+ * @author Paulo Cruz
+ */
+class FormContext @JvmOverloads constructor(
+	val isReadOnly: Boolean = false,
+	val isShowRequiredFieldsWarning: Boolean = false,
+	val isShowSubmitButton: Boolean = false,
```
Comment (victorg1991) [inline android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/FormContext.kt]: Can you apply the Javi suggestion here too? :)

File: android/library/src/main/java/com/liferay/mobile/screens/ddl/model/DDMStructure.java [hunk 6]
```diff
@@ -60,10 +64,33 @@ public DDMStructure createFromParcel(Parcel in) {
...
+        super();
+        this.name = name;
+        this.description = description;
+        this.pages = pages;
+        this.fields = CollectionsKt.flatMap(pages, new Function1<FormPage, Iterable<? extends Field>>() {
+            @Override
+            public Iterable<? extends Field> invoke(FormPage formPage) {
+                return formPage.getFields();
+            }
+        });
```
[the remark this answers is not in the document]
Comment (phcp) [reply android/library/src/main/java/com/liferay/mobile/screens/ddl/model/DDMStructure.java]: Yep. That's true, it's clearer using a for. (:

File: android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/Grid.kt [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+package com.liferay.mobile.screens.ddm.form.model
+
+import android.os.Parcel
+import android.os.Parcelable
+import java.io.Serializable
+
+/**
+ * @author Victor Oliveira
+ */
+class Grid @JvmOverloads constructor(
```
[the remark this answers is not in the document]
Comment (victorlaerte) [reply android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/Grid.kt]: Don't you think that GridElement sounds a little bit more like a row X column of the grid? @victorg1991

File: android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/FormContext.kt (hunk 5 above)
Comment (phcp) [reply android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/FormContext.kt]: Ok, I renamed on every place that had isShow. Thanks!

File: android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/Grid.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (victorg1991) [reply android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/Grid.kt]: I was wrong, I thought that it was that in fact: 
> a row X column of the grid

Omit my comment haha

**Comment by phcp** [comment]
Guys, all the changes are done, tell me if something is missing. About the 65k methods, I agree with @victorg1991, we should fix it in another PR. Thanks for the review! (:

**Comment by victorg1991** [comment]
Good job 🚀

**Pull request merged**